File: newver/services/xdelta_runner.py
from pathlib import Path
from PyQt6.QtCore import QObject, QProcess, pyqtSignal

class XDeltaRunner(QObject):
    output_received = pyqtSignal(str)
    finished = pyqtSignal(int)

    def __init__(self, executable: Path):
        super().__init__()
        self.process = QProcess(self)
        self.process.setProgram(str(executable))
        
        self.process.readyReadStandardOutput.connect(self._handle_stdout)
        self.process.readyReadStandardError.connect(self._handle_stderr)
        self.process.finished.connect(self._finished)

    def run(self, args: list[str]) -> None:
        self.process.setArguments(args)
        self.output_received.emit(f"Running: xdelta3 {' '.join(args)}\n")
        self.process.start()

    def is_running(self) -> bool:
        return self.process.state() == QProcess.ProcessState.Running

    def _handle_stdout(self):
        output = bytes(self.process.readAllStandardOutput()).decode('utf-8', errors='replace').strip()
        if output:
            self.output_received.emit(output)

    def _handle_stderr(self):
        error = bytes(self.process.readAllStandardError()).decode('utf-8', errors='replace').strip()
        if error:
            self.output_received.emit(f"ERROR: {error}")

    def _finished(self, exitCode: int, _):
        if exitCode == 0:
            self.output_received.emit("Process finished successfully.")
        else:
            self.output_received.emit(f"Process failed with exit code: {exitCode}")
        self.finished.emit(exitCode)

File: cpp/src/XDeltaRunner.cpp
#include "XDeltaRunner.h"

XDeltaRunner::XDeltaRunner(const QString &executable, QObject *parent)
    : QObject(parent)
    , process(new QProcess(this))
{
    process->setProgram(executable);
    connect(process, &QProcess::readyReadStandardOutput, this, &XDeltaRunner::handleStdout);
    connect(process, &QProcess::readyReadStandardError, this, &XDeltaRunner::handleStderr);
    connect(process, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
            this, &XDeltaRunner::onFinished);
}

void XDeltaRunner::run(const QStringList &args)
{
    process->setArguments(args);
    emit outputReceived(QString("Running: xdelta3 %1\n").arg(args.join(' ')));
    process->start();
}

bool XDeltaRunner::isRunning() const
{
    return process->state() == QProcess::Running;
}

void XDeltaRunner::handleStdout()
{
    QString out = QString::fromUtf8(process->readAllStandardOutput()).trimmed();
    if (!out.isEmpty())
        emit outputReceived(out);
}

void XDeltaRunner::handleStderr()
{
    QString err = QString::fromUtf8(process->readAllStandardError()).trimmed();
    if (!err.isEmpty())
        emit outputReceived("ERROR: " + err);
}

void XDeltaRunner::onFinished(int exitCode, QProcess::ExitStatus)
{
    if (exitCode == 0)
        emit outputReceived("Process finished successfully.");
    else
        emit outputReceived(QString("Process failed with exit code: %1").arg(exitCode));
    emit finished(exitCode);
}